### DISCUSSION TITLE
Wrapping Multiple Screen Under a Context Provider

### DISCUSSION THREAD

**Discussion by DVGY** [react-navigation/react-navigation#11977] [Q&A] [answered]
How can I wrap a screen under a Context Provider. I have seen all the issues and discussion and it seems there is no way to solve this ?

I am creating Screen based on Props being passed down. Now the problem here is that, I need to provide context for each screen and it seems that it is not possible.

Please help me here

**Comment by satya164** [comment, marked as answer]
React Navigation 7 supports a `layout` prop for `screen` component that can be used to add a wrapper (such as context provider) for the screen component.

https://reactnavigation.org/docs/7.x/screen#layout

**Reply by DVGY** [reply]
Ohh nice. I just have to wait for a stable release then.

Thanks

**Reply by Tronix117** [reply]
@satya164 
Any idea why the `screenLayout` is rendered outside of the provider defined within the `layout` ? at least on MaterialTopTabs 

Is it by design or is it an issue ?

**Reply by satya164** [reply]
@Tronix117 not sure what you mean. `screenLayout` is for specifying `layout` for multiple screens. If you specify `layout` for a screen, it will overwrite what's specified in `screenLayout` for that screen. both are not rendered together.

**Reply by Tronix117** [reply]
@satya164 sorry I was not clear, I'm not talking about the `screen.layout`, but I'm talking about the `navigator.layout`, like so:

```
createMaterialTopTabNavigator({
  layout: ({children}) => <ParentView><SomeProvider>{children}</SomeProvider></ParentView>,

  screenLayout: ({children}) => {
    const context = useSomeContext();
    return <SubView>{children}</SubView>
  },

  screens: { ... }
  ```

When doing that, ParentView is rendered around SubView => good
However it seems SubView is rendered first, so it does not have the context yet => bad

**Reply by satya164** [reply]
I'm unclear what you mean by SubView is rendered first. Rendered first before parent is rendered? Can you create a GitHub repo with code sample?

**Comment by jeremiebardon** [comment]
@satya164 I'm not sure it's a good practice to use screenLayout or layout for wrapping multiple screens in a provider. Since layout and screenLayout trigger a re-render each time route or navigation change I don't think it's a right place to put a provider, did I miss something ?

**Reply by Tronix117** [reply]
Agree with you, so I'm at the point I don't understand what is the use of Layout or ScreenLayout. When I define a layout I expect to use that to wrap my screens. So I can handle all animations between pages, add providers for this sub-navigation, ... So I expect no re-render... otherwise it means flat navigation is not usable for the majority of cases

**Reply by satya164** [reply]
> @satya164 I'm not sure it's a good practice to use screenLayout or layout for wrapping multiple screens in a provider. Since layout and screenLayout trigger a re-render each time route or navigation change I don't think it's a right place to put a provider, did I miss something ?

What does re-render have to do with providers?

Either way, something re-rendering shouldn't affect logic. If it does then there's something wrong with the logic.

**Reply by jeremiebardon** [reply]
He talks (I think) about the provider pattern more generally, this pattern could be used to provide initialisation and cleanup logic.

**Exemple:**
A user enters a conversation in a messaging app the conversation is fetch from the backend and loaded into the state.

This state is used to provide some utilities functions and state to all the screens inside my stack navigator through my provider (conversation, conversation thread, conversation information, conversation participants etc ...)

When the user leaves the conversation (the Stack navigator), the conversation state is cleanup.

It's also useful because it's not necessary to check and fetch the conversation on each screen since the user could come from push notification and can land on any screen of the navigator.

**The problem if we use screen layout on a stack navigator:** on EVERY  unloaded screen change in the stack the cleanup function of my provider is called.

To summarize: Using a custom provider in **screenLayout** for a Stack navigator is exactly the same as setting the provider inside layout on every stack screen in the navigator. It just avoids repetition for very simple use cases as outlined in the doc.

Could not be used for the case exposed above.

The solution for this could be wrapping the navigator with the provider (what I was doing before) but it's not very convenient since my provider needs a params (an ID) who comes from routes params

**Reply by jeremiebardon** [reply]
@satya164 you can correct me if I'm wrong but, at the end, the **screenLayout** is not intended to be used for this particular use case right ?

**Reply by satya164** [reply]
@jeremiebardon Yes, `layout` lets you wrap each screen in a wrapper. `screenLayout` is the same as `layout` - similar to how `screenOptions` lets you pass the same `options` to multiple screens or how `screenListeners` lets you pass the same listeners, `screenLayout` allows you to pass the same `layout` - otherwise it's functionally identical. Passing the same layout to multiple screens won't wrap all of those screens in a single wrapper.

If you need to keep a single state for all screens, you can't store that state inside the layout itself as you're specifying wrappers for each of the screens. However, you can store the state in the parent component and pass it down by using a context provider inside `layout` or `screenLayout`.

You can still trigger fetches, it just needs to store that state in the parent or a global cache.

In your example, how do you imagine it to work if it was a single wrapper? If your logic depends on the route params, and the wrapper wraps multiple screens - it can't have a single `route` object to read the params from. It's not possible to make it work with this API.

If you want a single wrapper, then you can use the [`layout`](https://reactnavigation.org/docs/navigator/#layout) prop on the navigator. There you have access to the navigator's state which contains all the rendered routes, you can read params for each route there and perform any logic such as fetching data.

Though something like React Query would simplify this kind of logic a lot.

**Reply by joprice** [reply]
@satya164 I'm trying to use a layout to propagate state to a group of routes, and it works well for the screen and headers, but not when I call setOptions to dynamically set the header. I have to pass it manually. Should components provided to params like `headerRight` dynamically be able to access the context that the layout provides?

**Reply by joprice** [reply]
I just noticed I'm using screenLayout in a group, so I'm guessing header is excluded from that. Is there a way to use layout with `if` to forward context only if a condition passes?

**Reply by satya164** [reply]
@joprice `screenLayout` only wraps screen content, so it's not possible to access them from components in options, as they can render at various levels. you could wrap your component in the context when calling `setOptions` as well, not the simplest, but I think there is no other way for now.

**Reply by satya164** [reply]
> Is there a way to use layout with if to forward context only if a condition passes?

Context will always get passed unconditionally, but you can always write a hook that consumes the context and conditionally return the value based on the route name, `useRoute` will work in header components.
